### PR TITLE
Add POST:/submissions and POST:/presubvalidation

### PR DESCRIPTION
**Pull request by lkuchenb**
@KerstenBreuer Haven't run this yet and your validation as well as the uniqueness validation is still missing, but it would be great of you could give feedback if you think this is okay conceptually.

### CODE REVIEW

File: datameta/api/submissions.py [hunk 1]
```diff
@@ -23,59 +23,143 @@
...
+    errors += [ (ref_fname, "Filename occurs multiple times in metadata") for ref_fname, count in ref_fname_counts.items() if count > 1 ]
+
+    # Make sure the files' filenames and the referenced filenames match
+    errors += [ (fname, "File included without reference in metadata") for fname in fnames.keys() if fname not in ref_fnames.keys() ]
+    errors += [ (ref_fname, "File referenced in metadata but not provided") for ref_fname in ref_fnames.keys() if ref_fname not in fnames ]
+
+    raise NotImplementedError()
+
+    # TODO
+    # - Run individual metadata validation. Q: We now have to associate the message not only with a field but also with a mset
```
Comment (KerstenBreuer) [inline datameta/api/submissions.py:96]: (Refering to 96) Yes that's true. If you don't like to include the mset ID in the "field", too, we could also add a "contextId" property to the ValidationError scheme?

**Review by KerstenBreuer — CHANGES_REQUESTED** [review]
Looks pretty perfect :)

File: datameta/api/submissions.py [hunk 2]
```diff
@@ -23,59 +23,143 @@
...
-
 @dataclass
 class SubmissionRequest(SubmissionBase):
     """SubmissionRequest container for OpenApi communication"""
-
-    def __json__(self, request: Request) -> dict:
-        
-        return self._json()
-
+    pass
```
[the remark this answers is not in the document]
Comment (lkuchenb) [reply datameta/api/submissions.py:40]: OK, will do

File: datameta/api/submissions.py [hunk 3]
```diff
@@ -23,59 +23,143 @@
...
-            }
-        )
-        return json_
+def validate_submission(request, auth_user):
+    db = request.dbsession
+
+    # Collect files
+    db_files = { file_id : resource.resource_by_id(db, models.File, file_id) for file_id in request.openapi_validated.body['fileIds'] }
+    # Collect missing files
+    errors = [ (file_id, "File not found") for file_id, db_file in db_files if db_file is None ]
```
[the remark this answers is not in the document]
Comment (lkuchenb) [reply datameta/api/submissions.py:53]: Indeed

**Comment by lkuchenb** [comment]
@KerstenBreuer If you find the time it would be great if you could have another look at this, it's feature complete from my side.

As for the error schema, I've extended it to have an `entity` member intended to be used for entity (resource) IDs. Thus, when there are multiple resources the response could refer to, we can use that to specify the resource as in

```json
{
"entity":"DMR-12345",
"field":"ZIP Code",
"message":"Invalid zip code"
}
```

Should this maybe be renamed to `resource` to remain in REST speak? 🤔

**Comment by KerstenBreuer** [comment]
> @KerstenBreuer If you find the time it would be great if you could have another look at this, it's feature complete from my side.
> 
> As for the error schema, I've extended it to have an `entity` member intended to be used for entity (resource) IDs. Thus, when there are multiple resources the response could refer to, we can use that to specify the resource as in
> 
> ```json
> {
> "entity":"DMR-12345",
> "field":"ZIP Code",
> "message":"Invalid zip code"
> }
> ```
> 
> Should this maybe be renamed to `resource` to remain in REST speak? 🤔

Great. I think both is fine and should be clear what it is referring to.

File: datameta/resource.py [hunk 4]
```diff
@@ -19,8 +19,28 @@
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 # SOFTWARE.
 
+from sqlalchemy import or_
 from uuid import UUID
 
+def resource_query_by_id(db, model, idstring):
```
Comment (KerstenBreuer) [inline datameta/resource.py:25]: Couldn't we marry the `resource_query_by_id` and `resource_by_id` functions? They seem redundant? Maybe one could use the `resource_query_by_id` inside the `resource_by_id`?

**Review by KerstenBreuer — APPROVED** [review]
Looks fine to me. Maybe just this one refactoring in resources.py?!

File: datameta/resource.py (hunk 4 above)
Comment (lkuchenb) [reply datameta/resource.py:25]: Done

**Pull request merged**